### PR TITLE
[RDY] Man and Help TOC

### PR DESCRIPTION
**Pull request by tweekmonster**
Adds a TOC using the location list for `:Man` and `:help`

### CODE REVIEW

File: runtime/autoload/man.vim [hunk 1]
```diff
@@ -190,6 +190,23 @@ function! s:open_page(sect, name)
...
+  call man#create_toc()
+endfunction
+
+function! man#create_toc() abort
+  if !exists('b:man_toc')
+    " Cache the TOC since user might switch in and out between the man pages.
+    let b:man_toc = []
+    for lnum in range(1, line('$'))
+      let text = getline(lnum)
+      if text =~# '^\%(\%(\S.*\)\=\S\|\%1l.*\|\s\{3\}\S.*\)$'
```
Comment (nhooyr) [inline runtime/autoload/man.vim]: I'm gonna update the syntax file to change the last regex there to instead of `\s`, just use ` `. So it looks like ` \{3\}\S.*`.

Comment (tweekmonster) [inline runtime/autoload/man.vim]: I originally used syntax for finding the lines because I was afraid of syntax changes, but it turned out to be pretty slow on large help files.  

Was there a problem with tabs being matched?

Comment (nhooyr) [inline runtime/autoload/man.vim]: No problems with tabs being matched but man sub headings are always prefixed with 3 spaces, not 3 whitespace characters. Just makes more sense.

**Comment by tweekmonster** [comment]
@nhooyr I have another potential PR after this one is merged:

```
autocmd BufEnter man://*
  \ if exists('b:man_toc') && line('$') == 1 && getline(1) =~# '^\s*$' |
  \   unlet! b:manwidth |
  \   execute 'silent Man' matchstr(expand('<amatch>'), 'man://\zs.*') |
  \ endif
```

That would allow the man page to reload if you run `:e` or restore from a session.  You see anything wrong with it?

**Comment by nhooyr** [comment]
@tweekmonster Great idea!
I'd change it to this

```vim
autocmd BufEnter man://*
      \ if exists('b:manwidth') && line('$') == 1 && getline(1) ==# '' |
      \   unlet b:manwidth |
      \   silent execute 'Man' matchstr(expand('<amatch>'), 'man://\zs.*') |
      \ endif
```

Instead of checking if `b:man_toc` exists, check if `b:manwidth` exists. Both variables work, but since we're already unsetting `b:manwidth` down there, it's more consistent. Also,`unlet` does not need to have a `!`.

I also think it is sufficient for `getline(1)` to just match an empty string. I also moved `silent` in front of `execute` but thats just personal preference because it gets syntax highlighted. Easier to recognize for me.

**Comment by jamessan** [comment]
> getline(1) ==# ''

Or `empty(getline(1))`.

**Comment by tweekmonster** [comment]
I always use `unlet!` for sanity.  Same reason for checking `line('$')` in the off chance a blank line is inserted at the top but the buffer has other content.

Actually, sanity is the same reason for the `silent` placement.  From my understanding things like `keepjumps`, `keepmarks`, `sandbox`, affect the command immediately to the left of it.  I don't think that's the case for `silent`, but whatevs.  I'm paranoid.

If it works for you without any issues for you, you should include the autocmd in your general fixes PR.  I tested `b:man_toc` since it's created after `filetype` is set and wouldn't have a chance of becoming recursive with the `:Man` command.  (again, paranoid).

**Comment by tweekmonster** [comment]
My mistake, I thought #5171 was misc/general fixes.

**Comment by nhooyr** [comment]
I included the autocmd in #5168.

**Comment by tweekmonster** [comment]
I think the only debatable thing about this PR is 4dd409e5d0f2094603e1411241e7ee62bc1919db.  Anyone have thoughts on changing `ftplugin/qf.vim` to hide the file name and line number in the location lists for the TOC?

**Comment by nhooyr** [comment]
I'm changing how `man.vim` works a bit in #5168 to make that autocmd work with sessions (`b:manwidth` is unset when restoring from a session) and address #5172. It will conflict with the changes here. Just giving you a heads up.

**Comment by nhooyr** [comment]
If #5249 is merged before this, I think this PR will need to be updated slightly again.

File: runtime/autoload/man.vim [hunk 2]
```diff
@@ -97,6 +97,23 @@ function! s:read_page(sect, name) abort
...
+  call man#create_toc()
+endfunction
+
+function! man#create_toc() abort
+  if !exists('b:man_toc')
+    " Cache the TOC since user might switch in and out between the man pages.
+    let b:man_toc = []
+    for lnum in range(1, line('$'))
+      let text = getline(lnum)
+      if text =~# '^\%(\%(\S.*\)\=\S\|\%1l.*\| \{3\}\S.*\)$'
```
Comment (nhooyr) [inline runtime/autoload/man.vim]: Nitpick but I think `\%1l.*` should be first in that regex.

Comment (nhooyr) [inline runtime/autoload/man.vim]: Actually, do we even need to add the title? Seems redundant. You can just press `G` if you want to go to the top. It'll always get you to the title because all the blank lines (if any) at the top are removed.

File: runtime/autoload/man.vim [hunk 3]
```diff
@@ -97,6 +97,23 @@ function! s:read_page(sect, name) abort
...
+        let text = substitute(getline(lnum), '\s\+', ' ', 'g')
```
Comment (nhooyr) [inline runtime/autoload/man.vim]: Why do we need this?

File: runtime/ftplugin/help.vim [hunk 4]
```diff
@@ -18,5 +18,45 @@ if has("conceal")
...
+      let add_text = ''
+      let text = getline(lnum)
+
+      if text =~# '^=\+$'
+        let text = getline(lnum + 1)
+        if text =~# '\*[^*]\+\*'
+          let lnum += 1
+          let add_text = matchstr(text, '.\{-}\ze\s*\*')
+        endif
+      elseif text =~# '^[A-Z][-A-Z .][-A-Z0-9 .()]*[ \t]\+\*'
```
Comment (nhooyr) [inline runtime/ftplugin/help.vim]: The regexp for `helpHeadline` starts off with `[-A-Z .]`. Why is `[A-Z]` in front?

File: runtime/ftplugin/help.vim [hunk 5]
```diff
@@ -18,5 +18,45 @@ if has("conceal")
...
+        let add_text = matchstr(text, '.\{-}\ze\s*\*')
+      elseif text =~# '^\u.*\s\+\~$'
+        let add_text = matchstr(text, '.\{-}\ze\s\+\~$')
+      elseif lnum == 1
```
Comment (nhooyr) [inline runtime/ftplugin/help.vim]: Same issue as the titles in manpages for this.

File: runtime/ftplugin/help.vim [hunk 6]
```diff
@@ -18,5 +18,45 @@ if has("conceal")
...
+        let add_text = text
+      endif
+
+      if !empty(add_text) && last_added != lnum
+        let last_added = lnum
+        call add(b:help_toc, {'bufnr': bufnr('%'), 'lnum': lnum,
+              \ 'filename': 'butt.txt',
```
Comment (nhooyr) [inline runtime/ftplugin/help.vim]: Why is that needed?

File: runtime/ftplugin/help.vim [hunk 7]
```diff
@@ -18,5 +18,45 @@ if has("conceal")
...
+              \ 'text': substitute(add_text, '\s\+', ' ', 'g')})
```
Comment (nhooyr) [inline runtime/ftplugin/help.vim]: Why do we need this?

File: runtime/ftplugin/qf.vim [hunk 8]
```diff
@@ -14,3 +14,10 @@ let b:undo_ftplugin = "set stl<"
 
 " Display the command that produced the list in the quickfix window:
 setlocal stl=%t%{exists('w:quickfix_title')?\ '\ '.w:quickfix_title\ :\ ''}\ %=%-15(%l,%c%V%)\ %P
+
+autocmd Syntax <buffer>
+      \ if exists('w:quickfix_title') && w:quickfix_title =~# '^:\%(Man\|Help\) TOC$' |
+      \   syntax match qfHideTOC "^.*| " conceal containedin=ALL |
+      \   setlocal concealcursor=nvi |
```
Comment (nhooyr) [inline runtime/ftplugin/qf.vim]: Why not set this to `nvic`? Doesn't make sense for the file name to pop up if I search the list.

File: runtime/ftplugin/qf.vim [hunk 9]
```diff
@@ -14,3 +14,10 @@ let b:undo_ftplugin = "set stl<"
...
+      \   setlocal conceallevel=2 |
```
Comment (nhooyr) [inline runtime/ftplugin/qf.vim]: Is there any reason why we use level 2 instead of 3 here?

**Comment by nhooyr** [comment]
I was testing this and I noticed some weird wrapping in the location list.

screen shot 2016-08-26 at 12 11 03 am

Not sure what's causing this.

File: runtime/autoload/man.vim (hunk 2 above)
Comment (tweekmonster) [inline runtime/autoload/man.vim]: It's mainly to have a complete TOC.  No other reason.

File: runtime/autoload/man.vim (hunk 3 above)
Comment (tweekmonster) [inline runtime/autoload/man.vim]: There were cases of right-aligned text or "table" column headers on some lines.  It would make the horizontal scrolling long.

File: runtime/ftplugin/help.vim (hunk 4 above)
Comment (tweekmonster) [inline runtime/ftplugin/help.vim]: Because some lines start with non-[A-Z] characters and I felt the ones that I saw weren't good TOC entries.

File: runtime/ftplugin/help.vim (hunk 6 above)
Comment (tweekmonster) [inline runtime/ftplugin/help.vim]: Hahahaha, I usually don't let these slip.  I use `butt`, `fart`, etc. as placeholder dummy testing text/vars because they're always unique in my sources.  I'll have to recheck.

File: runtime/ftplugin/qf.vim (hunk 9 above)
Comment (tweekmonster) [inline runtime/ftplugin/qf.vim]: `concealcursor=nvi` and `conceallevel=2` are my goto settings.  No other reason.

File: runtime/ftplugin/help.vim (hunk 6 above)
Comment (nhooyr) [inline runtime/ftplugin/help.vim]: Haha. I usually use `lol`, `foo`, `rekt` and some more as mine. 👍

**Comment by tweekmonster** [comment]
@nhooyr replied to your comments.  I'll revisit this when I find some time.

File: runtime/autoload/man.vim (hunk 3 above)
Comment (nhooyr) [inline runtime/autoload/man.vim]: IIRC, whitespace is trimmed in the location list automatically so that isn't necessary.

Comment (tweekmonster) [inline runtime/autoload/man.vim]: That definitely wasn't the case when I was testing this.

**Comment by nhooyr** [comment]
One more issue I found, you need a call to create the TOC when neovim is started as `$MANPAGER`. Just add it in the part of the ftplugin that checks if neovim is starting.

File: runtime/autoload/man.vim (hunk 2 above)
Comment (nhooyr) [inline runtime/autoload/man.vim]: Gotcha, we should also ignore the last line in a manpage. The date has no reason to be on the TOC.

File: runtime/autoload/man.vim (hunk 3 above)
Comment (nhooyr) [inline runtime/autoload/man.vim]: Yea, I edited my comment.

**Comment by nhooyr** [comment]
@tweekmonster may I continue this PR? I've got a lot of free time right now :P

**Comment by tweekmonster** [comment]
Oh for sure!  I feel bad that I forgot about it.  You should be able to pull from my fork to continue it, but it might be easier to start over since this is way outdated now.  Is there something I should do besides closing this PR?

Just a suggestion: Something I've been toying with recently is altering the [quickfix buffer to condense the location list](https://github.com/neomake/neomake/pull/325#issuecomment-268421805).  The filename is irrelevant and simply using syntax to conceal it is annoying because it still affects cursor movement.  The TOC might be a good place to do something like this.

**Comment by tweekmonster** [comment]
I'm going to pick this back up in the near future.

**Comment by tweekmonster** [comment]
Here's what it looks like now:

File: runtime/ftplugin/qf.vim [hunk 10]
```diff
@@ -14,3 +14,12 @@ let b:undo_ftplugin = "set stl<"
 
 " Display the command that produced the list in the quickfix window:
 setlocal stl=%t%{exists('w:quickfix_title')?\ '\ '.w:quickfix_title\ :\ ''}\ %=%-15(%l,%c%V%)\ %P
+
+autocmd Syntax <buffer>
+      \ if exists('w:quickfix_title') && w:quickfix_title =~# '^:\%(Man\|Help\) TOC$' && &syntax == 'qf' |
+      \   setlocal modifiable |
+      \   silent %delete _ |
+      \   call setline(1, map(deepcopy(getloclist(0)), 'v:val.text')) |
```
Comment (nhooyr) [inline runtime/ftplugin/qf.vim]: Is that `deepcopy` necessary there? If so, I am really surprised.

File: runtime/ftplugin/qf.vim [hunk 11]
```diff
@@ -14,3 +14,12 @@ let b:undo_ftplugin = "set stl<"
 
 " Display the command that produced the list in the quickfix window:
 setlocal stl=%t%{exists('w:quickfix_title')?\ '\ '.w:quickfix_title\ :\ ''}\ %=%-15(%l,%c%V%)\ %P
+
+autocmd Syntax <buffer>
```
Comment (nhooyr) [inline runtime/ftplugin/qf.vim]: What happens if I open a help window, open the location list. then switch to a man buffer. Will this autocmd still run and will the syntax change from `'help'` to `'man'`?

I'm not sure because you have `silent %delete _` in the command that is run, so I think the location list window is reused. And if so, then the syntax would be `'help'` and so the if statement would evaluate to false.

I don't have a laptop for a while, so I can't test this, just making sure.

File: runtime/ftplugin/help.vim [hunk 12]
```diff
@@ -18,5 +18,67 @@ if has("conceal")
...
+      if !empty(add_text) && last_added != lnum
+        let last_added = lnum
+        call add(b:help_toc, {'bufnr': bufnr('%'), 'lnum': lnum,
+              \ 'text': repeat('  ', level) . add_text})
+      endif
+      let lnum += 1
+    endwhile
+  endif
+
+  call setloclist(0, b:help_toc, ' ', 'Help TOC')
```
Comment (nhooyr) [inline runtime/ftplugin/help.vim]: Is it documented that the fourth argument to `setloclist` sets the title?

http://vimdoc.sourceforge.net/htmldoc/eval.html#setqflist()

Doesn't mention it.

**Review by nhooyr — CHANGES_REQUESTED** [review]
I reviewed just the man.vim changes for now.

File: runtime/ftplugin/help.vim (hunk 12 above)
Comment (jamessan) [inline runtime/ftplugin/help.vim]: The ability to set the title was implemented first in Neovim, and when it finally got merged upstream the 4th argument was changed to a dictionary with an optional `'title'` key.

I recently merged that upstream patch, and the documentation, but kept backwards compatibility with older Neovims by still recognizing a string as the 4th argument.

File: runtime/ftplugin/qf.vim (hunk 11 above)
Comment (tweekmonster) [inline runtime/ftplugin/qf.vim]: When the location list is updated, the window is reused, but the buffer isn't.  `filetype` gets set again and it starts over.

File: runtime/ftplugin/qf.vim (hunk 10 above)
Comment (tweekmonster) [inline runtime/ftplugin/qf.vim]: Heh, I'm not sure why that's there.  I was a little sleepy.

File: runtime/ftplugin/qf.vim (hunk 11 above)
Comment (nhooyr) [inline runtime/ftplugin/qf.vim]: Then what's with the ` silent %delete _` in the autocmd? If the buffer isn't reused, what is that deleting?

File: runtime/ftplugin/help.vim (hunk 12 above)
Comment (tweekmonster) [inline runtime/ftplugin/help.vim]: Based on what was in the docs at the time this was originally worked on: https://github.com/neovim/neovim/blob/v0.1.7/runtime/doc/eval.txt#L6135-L6187

Must've been updated to match Vim's.  It apparently still works, though.  I'll update it anyways.

**Comment by tweekmonster** [comment]
I'm thinking this should have a buffer `:TOC` command to generate the TOC.  For some of the longer help files, there's a noticeable pause.

File: runtime/ftplugin/qf.vim (hunk 11 above)
Comment (tweekmonster) [inline runtime/ftplugin/qf.vim]: The original lines set by Vim.  It felt cleaner than using syntax and conceal.  I secretly hate the way the location list renders.

File: runtime/ftplugin/help.vim (hunk 12 above)
Comment (tweekmonster) [inline runtime/ftplugin/help.vim]: @jamessan I am just now seeing your post 35 minutes after I posted.

**Comment by nhooyr** [comment]
If it's fast enough for man.vim, we should keep it automatic imo. Better user experience.

But then again, there is the argument of consistency. I'm not sure. 😕

**Review by nhooyr — APPROVED** [review]
man.vim changes look good.

**Comment by tweekmonster** [comment]
Actually just broke.  Apparently weird things are happening when setting the title the Right Way™

**Comment by tweekmonster** [comment]
> If it's fast enough for man.vim, we should keep it automatic imo. Better user experience.

Try `:help eval.txt` and `:Man tmux` to see what you think about the pause.  It's not that bad, but it'll be noticeable enough for people who hate new features that they personally don't find useful.

**Comment by justinmk** [comment]
It should be behind a command. We can think of a mapping too (`<M-]>` ?). Then mention this in the help files where the TOC normally lives.

    Type <M-]> to see table of contents.

File: runtime/ftplugin/help_toc.vim [hunk 13]
```diff
@@ -0,0 +1,66 @@
+if !exists('g:no_plugin_maps') && !exists('g:no_man_maps')
```
Comment (nhooyr) [inline runtime/ftplugin/help_toc.vim]: ;)

**Review by nhooyr — CHANGES_REQUESTED** [review]

File: runtime/ftplugin/help_toc.vim (hunk 13 above)
Comment (tweekmonster) [inline runtime/ftplugin/help_toc.vim]: Shhh

**Review by nhooyr — APPROVED** [review]

File: runtime/ftplugin/qf_toc.vim [hunk 14]
```diff
@@ -0,0 +1,23 @@
+function! s:setup_toc() abort
```
Comment (justinmk) [inline runtime/ftplugin/qf_toc.vim]: Hmm. The other location was better IMO. ftplugin/ should only contain filetypes. `qf_toc` isn't a filetype is it?

Comment (tweekmonster) [inline runtime/ftplugin/qf_toc.vim]: It's how `ftplugin.vim` loads them:

```vim
exe 'runtime! ftplugin/' . name . '.vim ftplugin/' . name . '_*.vim ftplugin/' . name . '/*.vim'
```

I did this for `help` and `qf` because I was adding substantial code and not completely rewriting them.  It just felt right.  I can put it back if you're not into it.

Comment (justinmk) [inline runtime/ftplugin/qf_toc.vim]: We've already committed to forking man.vim so no need to avoid it there. And with this PR we admit that we're forking the help ftplugin. So I'd say let's not dance around it.

File: runtime/ftplugin/help_toc.vim [hunk 15]
```diff
@@ -0,0 +1,66 @@
+if !exists('g:no_plugin_maps')
+  function! s:show_toc() abort
+    if !exists('b:qf_toc')
```
Comment (justinmk) [inline runtime/ftplugin/help_toc.vim]: Can we inspect the location list instead of storing stuff in a `b:` variable?

Comment (tweekmonster) [inline runtime/ftplugin/help_toc.vim]: I'm open to this, but how would I know that the location list is the TOC for *this* help file without recomputing to make the comparison?  I assume getting the first location list item's `bufnr`, but I don't like that since quickfix is an unreliable source for state information.

Comment (justinmk) [inline runtime/ftplugin/help_toc.vim]: Can the qf title be used? 

> I assume getting the first location list item's bufnr, but I don't like that since quickfix is an unreliable source for state information.

That sounds good too. Unreliable because the user could edit it? The worst that happens is the TOC is recalculated, so the stakes are low...

Comment (tweekmonster) [inline runtime/ftplugin/help_toc.vim]: > Can the qf title be used?

I guess `bufname()` could be attached to it, but that can be a little long and not exactly informative (in a useful way) since man and help should only have one window, and therefore one TOC.

Since the method now is to open the TOC when the key is pressed, I could just set a variable in the quickfix buffer that has the help/man's filename in it to track this.

> Unreliable because the user could edit it?

No.  A bit of a rant that's not directly related to this PR: It's just unreliable as a source of information regarding its relationship to another window.  Quickfix is meant to be programmatically filled, but setting the list is a one-way deal that strips away unknown dict keys that could've been used by a script to manage the list.  State info can't be attached until the quickfix window is opened, so, you can't just set the list then wait until the user has displayed it to script it further.

Also, the addition of the `{what}` argument to the quickfix functions is half-assed.

> The worst that happens is the TOC is recalculated

The worst case is bad UX.  I wanted to avoid the recalculation because it can be slow, and it also resets the quickfix selection index.  A `<M-]>` when the TOC is already opened for the current file should just move the cursor back to the TOC for browsing/searching.

**Comment by tweekmonster** [comment]
@justinmk I'm using the buffer name in a variable in the location list's window to keep track of what TOC it's displaying instead of keeping a list in the buffer.  Is that a more acceptable method?

File: runtime/autoload/man.vim [hunk 16]
```diff
@@ -152,23 +152,28 @@ function! s:put_page(page) abort
...
+  while lnum && lnum < last_line
+    let text = getline(lnum)
+    if text =~# '^\%( \{3\}\)\=\S.*$'
+      call add(toc, {'bufnr': bufnr('%'), 'lnum': lnum, 'text': text})
+    endif
+    let lnum = nextnonblank(lnum + 1)
+  endwhile
+
+  call setloclist(0, toc, ' ')
+  call setloclist(0, [], ' ', {'title': 'Man TOC'})
```
Comment (tweekmonster) [inline runtime/autoload/man.vim]: @jamessan Look, I'm eating my peas!

**Comment by tweekmonster** [comment]
@justinmk How do you feel about the change I made for tracking the current TOC?  I think it's ready for a squash and merge.

File: runtime/autoload/man.vim [hunk 17]
```diff
@@ -151,6 +151,31 @@ function! s:put_page(page) abort
...
+  lopen
+  let w:qf_toc = bufname
```
Comment (justinmk) [inline runtime/autoload/man.vim]: is this needed anymore?

File: runtime/doc/helphelp.txt [hunk 18]
```diff
@@ -25,6 +25,9 @@ Help on help files					*helphelp*
 			The 'helplang' option is used to select a language, if
 			the main help file is available in several languages.
 
+			Help file sections are added to the |location-list|
+			and can be viewed with |:lwindow|.
```
Comment (justinmk) [inline runtime/doc/helphelp.txt]: Are they added eagerly still? Or should this be changed to something like "To see the table of contents, type <M-]>."

File: runtime/ftplugin/help.vim [hunk 19]
```diff
@@ -18,5 +18,77 @@ if has("conceal")
   setlocal cole=2 cocu=nc
 endif
 
+if !exists('g:no_plugin_maps')
+  function! s:show_toc() abort
+    let bufname = bufname('%')
+    let info = getloclist(0, {'winid': 1})
+    if !empty(info) && getwinvar(info.winid, 'qf_toc') ==# bufname
```
Comment (justinmk) [inline runtime/ftplugin/help.vim]: Can we store the bufname in the getloclist `info` instead of `w:qf_toc`?

**Comment by justinmk** [comment]
The generic name for the "TOC" concept could be "outline". We could add this is a global concept. The idea being that different filetypes show outlines in different ways. For code files, an "outline" would usually show the current tags for a buffer. Eclipse uses the CTRL-O hotkey for this.

File: runtime/doc/helphelp.txt (hunk 18 above)
Comment (tweekmonster) [inline runtime/doc/helphelp.txt]: Oops! Yeah docs need to be updated.

File: runtime/ftplugin/help.vim (hunk 19 above)
Comment (tweekmonster) [inline runtime/ftplugin/help.vim]: How do you mean?  Arbitrary info can't be attached to qf lists (yet).

**Comment by tweekmonster** [comment]
> We could add this is a global concept. 

Dammit!  I actually started on a `autoload/toc.vim` as a thin wrapper to `setloclist()` that formatted the qf buffer.  I heard you in the back of my mind saying "ur over thinking this bro", so I scrapped it because I couldn't think of other uses in the debate I had with you in my mind.

**Comment by justinmk** [comment]
> Dammit! I actually started on a autoload/toc.vim as a thin wrapper to setloclist() that formatted the qf buffer.

Re "global concept", I mean having the concept be something that's provided differently in different filetypes. TOC is the particular application of "outline mode" for man and help buffers, but probably not for code files (I would guess tags/LSP/etc. would be used there instead).

So the "concept" is exposed as a mapping, I guess `<M-]>`. Similar to how `gf` works differently in different filetypes, but the concept is a core vim concept.

```
META-]        Show outline.
```

File: runtime/ftplugin/man.vim [hunk 20]
```diff
@@ -47,5 +48,7 @@ if get(g:, 'ft_man_folding_enable', 0)
   setlocal foldnestmax=1
 endif
 
+autocmd BufWinEnter <buffer> call man#create_toc()
```
Comment (justinmk) [inline runtime/ftplugin/man.vim]: Why is this still happening on BufWinEnter?

File: runtime/syntax/man.vim [hunk 21]
```diff
@@ -18,6 +18,11 @@ highlight default link manOptionDesc     Constant
 highlight default link manReference      PreProc
 highlight default link manSubHeading     Function
 
+if &filetype != 'man'
+  " Loaded by the location list TOC
+  finish
```
Comment (justinmk) [inline runtime/syntax/man.vim]: Is this because calling `man#show_toc()` triggers some other code that loads ft=man? Or why?

Comment (nhooyr) [inline runtime/syntax/man.vim]: Yes, it's for the syntax highlighting.
https://github.com/neovim/neovim/pull/5169/files#diff-cbbabc31237d18dd55a5545b0a9858f6R34

File: runtime/ftplugin/man.vim (hunk 20 above)
Comment (nhooyr) [inline runtime/ftplugin/man.vim]: That should definitely be removed, man#create_doc was removed.

File: runtime/syntax/man.vim (hunk 21 above)
Comment (justinmk) [inline runtime/syntax/man.vim]: This is for highlighting the qf window? That seems to really complicate the implementation. Is it really needed?

Comment (tweekmonster) [inline runtime/syntax/man.vim]: Is it *that* complicated, though?  It's not really doing anything specific for the qf window.  It's stopping it before executing script that's specific to the `man` plugin.  It actually makes the syntax loadable by other syntaxes.

Re: highlight in TOC, I'm a pretty strong proponent of notion that "color is information".  In Vim's help, the only defined sections are lines preceded by a `===` line.  Meaningful subsections either have all caps, or end with `~`.  There's some checks for the lines ending in `~` because it's sometimes used for table headings or code examples.  So, there's going to be false positives.  I felt that the highlighting made it more clear as to *why* a line is in the TOC in this case.

**Comment by tweekmonster** [comment]
> Re "global concept"

Ah okay.  So you're saying like an unofficial standard?  As in `<M-]>` should open up tagbar?

File: runtime/syntax/man.vim (hunk 21 above)
Comment (justinmk) [inline runtime/syntax/man.vim]: >  It actually makes the syntax loadable by other syntaxes.

Let's just change the comment then. Currently it specifically says "Loaded by the location list TOC" which is confusing.

    " syntax/man.vim may be included by other syntax files.

File: runtime/doc/filetype.txt [hunk 22]
```diff
@@ -544,6 +544,8 @@ CTRL-T                    Jump back to the location that the manpage was
                           opened from.
 q                         :quit if invoked as $MANPAGER, otherwise :close.
 
+To see the table of contents, type <M-]>.
```
Comment (tweekmonster) [inline runtime/doc/filetype.txt]: Any suggestions on how to display this.  The other keys are displayed as `CTRL-]` opposed to `<C-]>`. But,`META-]` and `ALT-]` doesn't look right and doesn't highlight.

**Comment by justinmk** [comment]
> Ah okay. So you're saying like an unofficial standard? As in <M-]> should open up tagbar?

Yes. Not in this PR, but eventually. It'll be just as official as `<C-]>`, `:help include-search`, and all other built-in concepts :)

File: runtime/doc/filetype.txt (hunk 22 above)
Comment (justinmk) [inline runtime/doc/filetype.txt]: We should make `META-foo` and `ALT-foo` highlight. In `runtime/syntax/help.vim`:

```vim
syn match helpSpecial>  >       "CTRL-."
syn match helpSpecial>  >       "META-."
syn match helpSpecial>  >       "ALT-."
```

Comment (tweekmonster) [inline runtime/doc/filetype.txt]: Which is more accurate/helpful to use in the docs?  `META-` or `ALT-`?  I feel like `ALT-` should be used because it's actually on the keyboard.  But, I don't know enough to know if `META-` is always the same as `ALT-`.  My vague recollection is that they aren't always the same.

Comment (tweekmonster) [inline runtime/doc/filetype.txt]: Also, should I go ahead and add highlights for `META-Break`, etc?

Comment (justinmk) [inline runtime/doc/filetype.txt]: For the man.vim case, this should be listed like the other mappings above.

    META-]                           Show the manpage outline in the |location-list|.

File: runtime/ftplugin/qf.vim [hunk 23]
```diff
@@ -14,3 +14,28 @@ let b:undo_ftplugin = "set stl<"
...
+
+  let bufnr = list[0].bufnr
+  setlocal modifiable
+  silent %delete _
+  call setline(1, map(list, 'v:val.text'))
+  setlocal nomodifiable nomodified
+  let &syntax = getbufvar(bufnr, '&syntax')
+endfunction
+
+
```
Comment (justinmk) [inline runtime/ftplugin/qf.vim]: double newlines is inconsistent with the rest of the project

Comment (tweekmonster) [inline runtime/ftplugin/qf.vim]: It's a habit I have from adhering to PEP8.  I'll fix it.

File: runtime/doc/filetype.txt (hunk 22 above)
Comment (justinmk) [inline runtime/doc/filetype.txt]: > Also, should I go ahead and add highlights for META-Break, etc?

Up to you, I'd say wait until we actually need it.

**Comment by tweekmonster** [comment]
@justinmk It'd be cool if this made it into the release.  Is there something outstanding that I'm forgetting?

**Comment by justinmk** [comment]
I was hemming and hawing about the mapping. I suppose we can change it later..

**Comment by justinmk** [comment]
@tweekmonster I pulled this down locally, the location list is coming up empty in help pages and man pages:

```
nvim -u NORC
:h h
<M-]>
```

Are you seeing differently?

**Comment by tweekmonster** [comment]
@justinmk Yeah it's apparently coming up blank now.  Investigating.

**Comment by tweekmonster** [comment]
@justinmk It's fixed now.  Apparently setting `{what}` follows the rules of `{action}` now for some great reason.

**Comment by justinmk** [comment]
Merged in https://github.com/neovim/neovim/pull/6632. Thanks @tweekmonster !
